### PR TITLE
Update ClearML Integration

### PR DESCRIPTION
**Pull request by thepycoder**
This PR adds better ClearML integration (based on the relevant changes recently made in YOLOv5 too)

Some open questions though:
- [ ] There seems to be no clean way to get to the images and their predictions, to visualize some samples like in YOLOv5. There are the mosaics, but sometimes getting singular images is nice! (Also need some config settings for that)
- [ ] Is there a good way to implement the dataset integration like in YOLOv5? So users can provide a `clearml://` URL.
- [ ] How to deal with latest models? There is no config setting for save interval that I can see?

If these are features or interfaces actively being worked on, great! Otherwise let's discuss them and see what I can/cannot add :)

## 🛠️ PR Summary

<sub>Made with ❤️ by [Ultralytics Actions](https://github.com/ultralytics/actions)<sub>

### 🌟 Summary
- N/A (No changes to summarize)

### 📊 Key Changes
- None (The PR contains no code changes)

### 🎯 Purpose & Impact
- N/A (Without any changes, there's no impact to elucidate)

### CODE REVIEW

**Comment by thepycoder** [comment]
I have read the CLA Document and I sign the CLA

**Comment by thepycoder** [comment]
@AyushExel Do you have any update on the questions posed here? If not, it still makes sense to go forward with the PR, since it does add some nice extra features :)

**Comment by AyushExel** [comment]
Hey sorry I missed the tag on this. Actually I don't really know the answers just yet. We'll need to discuss this internally. Let's touch base next week

**Comment by thepycoder** [comment]
Hey @AyushExel any updates on this? :) We can always handle the questions separately from the PR itself if you want!

**Comment by AyushExel** [comment]
@thepycoder Hey I completely missed this mention. Yes, why don't you open an email thread bw me, you, and Glenn. I think we'll need to meet to finalize this

**Comment by glenn-jocher** [comment]
@thepycoder hi there! Apologies for the delay on this, this PR seems to have slipped below our radar. I've added a TODO to prioritize integration. 

I just resolved conflicts and merged main into this branch. Since it's been a while since submission can you test this PR and verify/update everything, and let us know when it's good to merge?

Thank you!

**Comment by glenn-jocher** [comment]
@thepycoder about your last question, best.pt and last.pt are saved (and overwritten if applicable) every epoch by default. There is an argument called save_period that additionally allows for saving checkpoints on a period basis that are never overwritten, i.e. epoch_10.pt, epoch_20.pt for `save_period=10`

File: ultralytics/yolo/utils/callbacks/clearml.py [hunk 1]
```diff
@@ -12,49 +17,104 @@
...
-        _log_images({f.stem: str(f) for f in trainer.save_dir.glob('train_batch*.jpg')}, 'Mosaic', trainer.epoch)
+        _log_debug_samples(sorted(trainer.save_dir.glob('train_batch*.jpg')), 'Mosaic')
 
 
 def on_fit_epoch_end(trainer):
-    task = Task.current_task()
-    if task and trainer.epoch == 0:
+    # You should have access to the validation bboxes under jdict
+    Task.current_task().get_logger().report_scalar('Epoch Time',
+                                                   'Epoch Time',
```
Comment (glenn-jocher) [inline ultralytics/yolo/utils/callbacks/clearml.py:84]: @thepycoder is this duplicated input correct?

**Comment by glenn-jocher** [comment]
@thepycoder you can pre-commit errors in CI that are unresolved also. Please resolve:

Screenshot 2023-04-04 at 13 47 37

File: ultralytics/yolo/utils/callbacks/clearml.py (hunk 1 above)
Comment (thepycoder) [reply ultralytics/yolo/utils/callbacks/clearml.py:84]: Yes it is, it is meant as a title and a series name. In this case there is only 1 series on the graph so I tend to call them the same. Series can also be an empty string, but it is a required parameter.

I'm updating the code with keyword arguments, so it's at least a little more readable :)

**Comment by glenn-jocher** [comment]
@thepycoder great to hear that! If you encounter any issues, feel free to bring them up and reach out for support. It's important for us to ensure that everything works well before merging any changes to the codebase. Once everything checks out, we can discuss merging the changes into the main branch.

**Comment by thepycoder** [comment]
@glenn-jocher This looks good to me now :)

**Comment by glenn-jocher** [comment]
@thepycoder thanks for the updates! This looks good, but can you expand the list-comprehension into actual for-loops where appropriate? I think it's generally not best practices to create lists that don't return as you have in a few places. 

Once you've done this I'll go ahead and merge. Thanks!

**Comment by thepycoder** [comment]
Done!

**Comment by glenn-jocher** [comment]
@thepycoder looks good! Merged PR into updates branch which will publish to 8.0.67 soon.

**Comment by glenn-jocher** [comment]
@thepycoder please take a look at new ClearML bug introduced by this PR that users are seeing now in https://github.com/ultralytics/ultralytics/issues/1882

I'd recommend re-introducing the code we had in place in each callback before this PR that checked if a task existed before doing anything:

```
task = Task.current_task()
if task:
    # RUN TASKS
```

**Pull request merged**